File: 4.ConvertStringToLowercase/PYTHON/main.py
from ConvertStringToLowercase import ConvertStringToLowercase
import json
import os

class Util:
    @staticmethod
    def read_json(file_path):
        with open(file_path, 'r') as file:
            return json.load(file)

    @staticmethod
    def write_json(file_path, data):
        with open(file_path, 'w') as file:
            json.dump(data, file, indent=4)

class ConvertStringToLowercaseTest:
    def __init__(self):
        self.util = Util()
        self.config = self.util.read_json('conf/config.json')
        self.testcases = self.util.read_json(self.config.get('testcasesFile'))

    def test(self, input_string, expected_output):
        # Test the ConvertStringToLowercase class here
        converter = ConvertStringToLowercase()
        result = converter.convert_to_lowercase(input_string)
        return result == expected_output

def main():
    tester = ConvertStringToLowercaseTest()
    testcases_json = tester.testcases

    for testcase in testcases_json:
        res = tester.test(testcase['input'], testcase['output'])
        testcase["result"] = "fail" if not res else "pass"

    # Update the result in python_result.json
    tester.util.write_json(tester.config.get('PythonResultFile'), testcases_json)

    if "fail" in [testcase["result"] for testcase in testcases_json]:
        raise Exception("Testcases failed")
    print("All testcases passed")

if __name__ == "__main__":
    main()

File: 4.ConvertStringToLowercase/CPP/src/main.cpp
#include "../include/json.hpp"
#include "../include/ConvertStringToLowercase.hpp"
#include <fstream>
#include <iostream>
#include <vector>
#include <string>

using json = nlohmann::json;

class Config {
public:
    std::string testcasesFile;
    std::string CppResultFile;
};

class TestCase {
public:
    std::string input;
    std::string output;
    bool result;

    void to_json(json& j) {
        j = json{{"input", input}, {"output", output}, {"result", result}};
    }
};

json serialize(std::vector<TestCase> testcases) {
    json outputJson = json::array();
    for (TestCase testcase : testcases) {
        json j;
        testcase.to_json(j);
        outputJson.push_back(j);
    }
    return outputJson;
}

class Util {
public:
    std::vector<TestCase> testcases;
    Config config;

    Util() {
        config = readConfig();
        testcases = readTestcases();
    }

    Config readConfig() {
        std::ifstream configFile;
        configFile.open("../../conf/config.json");
        json configJson;
        configFile >> configJson;
        configFile.close();
        Config config;
        config.testcasesFile = configJson["testcasesFile"];
        config.CppResultFile = configJson["CppResultFile"];
        return config;
    }

    std::vector<TestCase> readTestcases() {
        std::ifstream testcaseFile;
        testcaseFile.open("../../" + config.testcasesFile);
        json j;
        testcaseFile >> j;
        testcaseFile.close();
        std::vector<TestCase> testcases;
        for (int i = 0; i < j.size(); i++) {
            TestCase testcase;
            testcase.input = j[i]["input"].get<std::string>();
            testcase.output = j[i]["output"].get<std::string>();
            testcases.push_back(testcase);
        }
        return testcases;
    }

    void writeResult() {
        std::ofstream resultFile;
        resultFile.open("../../" + config.CppResultFile);
        json j = serialize(testcases);
        resultFile << j.dump(2);
        resultFile.close();
    }
};

int main() {
    Util util;
    ConvertStringToLowercase converter;
    std::vector<TestCase>& testcases = util.testcases;
    bool failed = false;
    for (TestCase& testcase : testcases) {
        std::string res = converter.convertToLowercase(testcase.input);
        if (res != testcase.output) {
            failed = true;
        }
        testcase.result = !failed;
    }
    util.writeResult();
    if (failed) {
        throw std::runtime_error("Some test cases failed");
    }
    std::cout << "All test cases passed" << std::endl;
    return 0;
}
